### PR TITLE
[Sampling] Implement `dgl.to_block()` for the GPU

### PR DESCRIPTION
**Pull request by nv-dlasalle**
## Description
This enables `dgl.to_block()` to be called on frontiers that have already been transferred to the GPU.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

## Changes

This adds the files `src/graph/transform/cuda/cuda_to_block.h` and `src/graph/transform/cuda/cuda_to_block.cu`, and integrates the new function in `src/graph/transform/to_bipartite.cc`, as well as modifications to `python/dgl/transform.py` to enable the GPU version.

The unit tests for `to_block()` in `tests/compute/test_transform.py` have been enabled for the GPU.

Note, this is far from an optimal version as it processes each edge type independently, and the kernels for generating the re-numbering to match the sequential CPU code has room for improvement. In the current matser branch, the time is dominated by `cudaMalloc()` and `cudaFree()`:
to_block_gpu_with_allocs

For RGCN on ogbn-mag with GPU memory pooling enabled (either via #2328 or #2263) to avoid the `cudaMalloc()`/`cudaFree()` calls, it's about an order of magnitude faster than the CPU overall, and about 20x faster when just looking at the C++ function.

to_block_gpu

### CODE REVIEW

**Comment by BarclayII** [comment]
@jermainewang I have a question on how to use the GPU version of `to_block`.

Right now our data loader alternates between

* Calling `sample_neighbors` from a set of seed nodes.
* Calling `to_block` to obtain the bipartite graph as well as the seed nodes for the previous GNN layer.

Since we only have `sample_neighbors` on CPU, how would we use GPU's `to_block`?  Put the result from `sample_neighbors`, call `to_block`, then put the seed nodes for the previous GNN layer back to CPU for the next `sample_neighbor` call?  If that's the case, then DataLoaders cannot work with multiprocessing?

I intend to extend the DataLoader with a `device` argument so that the stuff returned by DataLoader is already on GPU, but the problem comes into my way.

**Comment by nv-dlasalle** [comment]
@BarclayII I think the idea, is that the dataloading processes would just perform sampling, and generate the unique set of dst nodes per layer on the CPU. The subgraphs from sampling would then be sent to the training process, which would copy them up to the GPU, where then `to_block()` would be invoked (@jermainewang and @zheng-da correct me if I'm wrong on this).

Alternatively, this would also support the future case where sampling is done on the GPU (either in the small case where the graph fits in one GPU, or the graph is partitioned across multiple GPUs).

File: src/graph/transform/cuda/cuda_to_block.cu [hunk 1]
```diff
@@ -0,0 +1,972 @@
...
+
+  /**
+  * @brief This function builds node maps for each node type, preserving the
+  * order of the input nodes.
+  *
+  * @param nodes The set of input nodes.
+  * @param nodes_maps The constructed node maps.
+  * @param workspace The scratch space to use.
+  * @param workspaceSize The size of the scratch space.
+  * @param stream The stream to operate on.
```
Comment (jermainewang) [inline src/graph/transform/cuda/cuda_to_block.cu:598]: The docstring misses some of the arguments.

**Comment by jermainewang** [comment]
@nv-dlasalle Sorry for the extremely late review. I scanned the code the main logic looks fine for me. The `cuda_to_block.cu` is a little bit lengthy, so I suggest break it down into multiple files and have file-level docstring to describe them.
* `AtomicCAS` related functions are useful elsewhere too. Consider putting them to `src/runtime/cuda`.
* The GPU hashmap can have a dedicated `.cuh` file and it will be great if you could give some descriptions on how to use it. I think many functions are for internal implementation so I'd like to have an clean interface for developers who just want to use them.
* It will be great if you could add more descriptions to `generate_hashmap_unique` and `generate_hashmap_duplicate` such as what are the recommended practices?

Thanks always for the great effort!

File: python/dgl/transform.py [hunk 2]
```diff
@@ -1723,10 +1724,15 @@ def to_block(g, dst_nodes=None, include_dst_in_src=True):
...
     dst_node_ids = [
-        utils.toindex(dst_nodes.get(ntype, []), g._idtype_str).tousertensor()
+        utils.toindex(dst_nodes.get(ntype, []), g._idtype_str).tousertensor(
+            ctx=F.to_backend_ctx(g._graph.ctx))
         for ntype in g.ntypes]
     dst_node_ids_nd = [F.to_dgl_nd(nodes) for nodes in dst_node_ids]
 
+    for d in dst_node_ids_nd:
+        if g._graph.ctx != d.ctx:
+            raise ValueError('g and dst_nodes need to have the same context.')
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply python/dgl/transform.py:1743]: Move into the C++ code in d4b709c

File: python/dgl/transform.py [hunk 3]
```diff
@@ -1629,6 +1629,9 @@ def to_block(g, dst_nodes=None, include_dst_in_src=True):
     DGLError
         If :attr:`dst_nodes` is specified but it is not a superset of all the nodes that
         have at least one inbound edge.
+    ValueError
+        If :attr:`dst_nodes` is not None, and :attr:`g` and :attr:`dst_nodes`
+        are not in the same context.
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply python/dgl/transform.py:1637]: Moved into the C++ code in 
d4b709c , as well as changed the pre-existing check from raising a valueerror, to raising a DGLError.

File: src/graph/transform/to_bipartite.cc [hunk 4]
```diff
@@ -127,8 +128,18 @@ DGL_REGISTER_GLOBAL("transform._CAPI_DGLToBlock")
...
+#ifdef DGL_USE_CUDA
+      std::tie(new_graph, lhs_nodes, induced_edges) = cuda::CudaToBlock(
+          graph_ref.sptr(), rhs_nodes, include_rhs_in_lhs);
+#else
+      CHECK(false) << "GPU support not enabled.";
+#endif
+    } else {
+      std::tie(new_graph, lhs_nodes, induced_edges) = ToBlock(
+          graph_ref.sptr(), rhs_nodes, include_rhs_in_lhs);
+    }
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/graph/transform/to_bipartite.cc]: Changed to using macro in 517ac46

File: src/graph/transform/cuda/cuda_to_block.cu (hunk 1 above)
Comment (nv-dlasalle) [reply src/graph/transform/cuda/cuda_to_block.cu]: Updated in 
85196a2

**Comment by nv-dlasalle** [comment]
@jermainewang I split the hashtable functionality out into `/src/runtime/cuda/cuda_hashtable.cu` and `/src/runtime/cuda/cuda_hashtable.cuh`. I looked into trying to make the hashtable more generic, but given the need to track order of insertions, and to maintain performance, I think its best to leave it specialized for now. I've named it `OrderedHashTable`, in the `dgl::runtime::cuda` namespace. Let me know if you want any changes there. Thanks!

**Review by jermainewang — APPROVED** [review]

File: src/runtime/cuda/cuda_device_common.cuh [hunk 5]
```diff
@@ -0,0 +1,73 @@
+/*!
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/runtime/cuda/cuda_device_common.cuh]: I moved the AtomicCAS wrappers to `/src/kernel/cuda/atomic.cuh` in 7069546 , as I had missed that file before.

File: src/runtime/cuda/cuda_hashtable.cuh [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+  }
+};
+
+
+
+/**
+* \brief An on GPU hash table implementation, which preserves the order of
+* inserted elements. It uses quadratic probing for collisions. The size/number
+* buckets the hash table is created with, should be much greater than the
+* number of unique items to be inserted. Using less than the number of
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/runtime/cuda/cuda_hashtable.cuh]: I expand the explanation in bda1493, as well as added example usage for filling the hashtable, and using it to map indexes.

File: src/runtime/cuda/cuda_hashtable.cuh [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+      IdType local;
+      int64_t index;
+    };
+
+    static constexpr int DEFAULT_SCALE = 3;
+
+    typedef Mapping* Iterator;
+
+    // Must be uniform bytes for memset to work
+    static constexpr IdType EmptyKey = static_cast<IdType>(-1);
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/runtime/cuda/cuda_hashtable.cuh]: Moved in f08b370

File: src/runtime/cuda/cuda_hashtable.cuh [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+* number of unique items to be inserted. Using less than the number of
+*/
+template<typename IdType>
+class OrderedHashTable {
+  public:
+    struct Mapping {
+      IdType key;
+      IdType local;
+      int64_t index;
+    };
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/runtime/cuda/cuda_hashtable.cuh:85]: Documented in 2dc479a

File: src/runtime/cuda/cuda_hashtable.cuh [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+#include "cuda_runtime.h"
+#include "cuda_common.h"
+#include "cuda_device_common.cuh"
+
+namespace dgl {
+namespace runtime {
+namespace cuda {
+
+template<typename IdType>
+struct BlockPrefixCallbackOp {
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/runtime/cuda/cuda_hashtable.cuh]: Moved in f08b370

File: src/runtime/cuda/cuda_hashtable.cuh [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+
+    void FillWithUnique(
+        const IdType * const input,
+        const size_t num_input,
+        DGLContext ctx,
+        cudaStream_t stream);
+
+    inline __device__ Iterator Insert(
+        const IdType id,
+        const size_t index);
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/runtime/cuda/cuda_hashtable.cuh]: I added doxygen comments in bda1493, and refactored it to hide the `Insert()` methods from the public interface, as well as added use case examples.

File: src/runtime/cuda/cuda_hashtable.cu [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+* \brief This counts the number of nodes inserted per thread block.
+*
+* \tparam IdType The type of of id.
+* \tparam BLOCK_SIZE The size of the thread block.
+* \tparam TILE_SIZE The number of entries each thread block will process.
+* \param nodes The nodes ot insert.
+* \param num_nodes The number of nodes to insert.
+* \param hash_size The size of the hash table.
+* \param pairs The hash table.
+* \param num_unique_nodes The number of nodes inserted into the hash table per thread
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/runtime/cuda/cuda_hashtable.cu]: Fixed in 
c92ee24

File: src/runtime/cuda/cuda_hashtable.cuh [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+    static constexpr IdType EmptyKey = static_cast<IdType>(-1);
+
+    /**
+    * \brief Get the amount of workspace the hashtable needs to function, given
+    * the number of positions.
+    *
+    * \param size The number of items to insert into the hash table.
+    * \param scale The power of two times larger the number of buckets should
+    * be than the number of items. Increasing should decrease the number of
+    * collisions in the hashtable, at the expense of memory consumption.
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/runtime/cuda/cuda_hashtable.cuh]: In refactoring this got removed in f08b370, but in the constructor (where size and scale are now input), I've added documentation to describe the memory usage as `O(size*2^scale)`.

File: src/runtime/cuda/cuda_hashtable.cuh [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+      IdType pos = Hash(id);
+
+      // linearly scan for matching entry
+      IdType delta = 1;
+      while (table_[pos].key != id) {
+        assert(table_[pos].key != EmptyKey);
+        pos = Hash(pos+delta);
+        delta +=1;
+      }
+      assert(pos < size_);
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/runtime/cuda/cuda_hashtable.cuh:149]: Next cudaError_t returned (likely from synchornizing) will be cudaErrorAssert (710). 

From cuda-gdb you can get information like this:
```
[Switching focus to CUDA kernel 0, grid 12, block (0,0,0), thread (160,0,0), device 0, sm 0, warp 5, lane 0]
0x0000555556ff2cd0 in void nvcomp::(anonymous namespace)::deferredCopy<unsigned char, 512>(unsigned char**, unsigned char const*, unsigned long const*)<<<(1,1,1),(512,1,1)>>> ()
```

**Pull request merged**